### PR TITLE
Improve diagnostic messages during learn

### PR DESCRIPTION
**Pull request by Sopel97**
This PR improves the diagnostics during learn by the following means:

- unification of the message format
- a `verbose` flag for `learn` that can be used to enable more granular diagnostics
- in `update_parameters` when `verbose == true` there is additional output regarding the gradients and the differences between evaluation of the discretized net and the trained net (they should be relatively close, otherwise things are wrong)
- display of layers and their indices in forward order before learning.
- display of layer name for which `check_health` is performed
- `check_health` for affine transform layer with average biases, weights, and last changes to them
- `check_health` in clipped relu and feature transformer shows the % of clipped outputs
- `check_health` in the feature transformer shows the average abs bias and weight
- output is synchronized in regions in a non-blocking manner. So for example during calc_loss all outputs will be seen on the console without interleaving from other sources (like sfen reader), but the other sources will not need to wait for calc_loss completion during io.
- all `learn` parameters are printed

Sample output with `verbose`
```
Stockfish 241020 by the Stockfish developers (see AUTHORS file)
info string Loaded eval file nn-98a7585c85e9.nnue
setoption name SkipLoadingEval value true
setoption name Threads value 4
setoption name Use NNUE value pure
learn targetdir data_q epochs 400 batchsize 100000 use_draw_in_training 1 use_d
aw_in_validation 1 lr 1 lambda 1 eval_limit 32000 nn_batch_size 1000 newbob_dec
y 0.99 eval_save_interval 10000000 loss_output_interval 1000000 set_recommended
uci_options verbose
INFO: Executing learn command
INFO: Input files:
  - data_q/10m_d3_2.binpack
INFO: Parameters:
  - epochs                   : 400
  - epochs * minibatch size  : 40000000
  - eval_limit               : 32000
  - save_only_once           : false
  - shuffle on read          : true
  - Loss Function            : ELMO_METHOD(WCSC27)
  - minibatch size           : 100000
  - nn_batch_size            : 1000
  - nn_options               :
  - learning rate            : 1
  - use draws in training    : 1
  - use draws in validation  : 1
  - skip repeated positions  : 1
  - winning prob coeff       : 0.00276753
  - use_wdl                  : 0
  - src_score_min_value      : 0
  - src_score_max_value      : 1
  - dest_score_min_value     : 0
  - dest_score_max_value     : 1
  - reduction_gameply        : 1
  - LAMBDA                   : 1
  - LAMBDA2                  : 1
  - LAMBDA_LIMIT             : 32000
  - eval_save_interval       : 10000000 sfens
  - loss_output_interval     : 1000000 sfens
  - sfen_read_size           : 10000000
  - thread_buffer_size       : 10000
  - seed                     :
  - verbose                  : true
  - learning rate scheduling : newbob with decay
  - newbob_decay             : 0.99
  - newbob_num_trials        : 4

INFO: Started initialization.
INFO (initialize_training): Initializing NN training for Features=HalfKP(Friend
[41024->256x2],Network=AffineTransform[1<-32](ClippedReLU[32](AffineTransform[3
<-32](ClippedReLU[32](AffineTransform[32<-512](InputSlice[512(0:512)])))))

Layers:
  - 0 - HalfKP(Friend)[41024->256x2]
  - 1 - InputSlice[512(0:512)]
  - 2 - AffineTransform[32<-512]
  - 3 - ClippedReLU[32]
  - 4 - AffineTransform[32<-32]
  - 5 - ClippedReLU[32]
  - 6 - AffineTransform[1<-32]

INFO (initialize_training): Performing random net initialization.
Finished initialization.
info string NNUE evaluation using  enabled
INFO (sfen_reader): Opened file for reading: data_q/10m_d3_2.binpack

PROGRESS (calc_loss): Sun Oct 25 11:41:12 2020, 0 sfens, 0 sfens/second, epoch

  - learning rate = 1
  - startpos eval = 0
  - test_cross_entropy_eval = 0.693145
  - test_cross_entropy_win  = 0.693145
  - test_entropy_eval       = 0.242816
  - test_entropy_win        = 0.17987
  - test_cross_entropy      = 0.693145
  - test_entropy            = 0.242816
  - norm = 0
  - move accuracy = 8.2%
INFO (sfen_reader): Opened file for reading: data_q/10m_d3_2.binpack
INFO (learn): initial loss = 0.450329
INFO (update_parameters): epoch = 1 , avg_abs(trainer_eval-nnue_eval) = 0 , avg
abs(nnue_eval) = 0 , avg_relative_error = nan , batch_size = 1000 , grad_norm =
398.662
INFO (update_parameters): epoch = 2 , avg_abs(trainer_eval-nnue_eval) = 2.249 ,
avg_abs(nnue_eval) = 54.561 , avg_relative_error = 0.0412199 , batch_size = 100
 , grad_norm = 377.229
INFO (update_parameters): epoch = 3 , avg_abs(trainer_eval-nnue_eval) = 5.958 ,
avg_abs(nnue_eval) = 22.385 , avg_relative_error = 0.26616 , batch_size = 1000
 grad_norm = 387.942
INFO (update_parameters): epoch = 4 , avg_abs(trainer_eval-nnue_eval) = 0 , avg
abs(nnue_eval) = 85 , avg_relative_error = 0 , batch_size = 1000 , grad_norm =
85.77
INFO (update_parameters): epoch = 5 , avg_abs(trainer_eval-nnue_eval) = 0 , avg
abs(nnue_eval) = 89 , avg_relative_error = 0 , batch_size = 1000 , grad_norm =
83.685
INFO (update_parameters): epoch = 6 , avg_abs(trainer_eval-nnue_eval) = 1 , avg
abs(nnue_eval) = 66 , avg_relative_error = 0.0151515 , batch_size = 1000 , grad
norm = 386.288
INFO (update_parameters): epoch = 7 , avg_abs(trainer_eval-nnue_eval) = 1 , avg
abs(nnue_eval) = 90 , avg_relative_error = 0.0111111 , batch_size = 1000 , grad
norm = 387.619
INFO (update_parameters): epoch = 8 , avg_abs(trainer_eval-nnue_eval) = 1 , avg
abs(nnue_eval) = 112 , avg_relative_error = 0.00892857 , batch_size = 1000 , gr
d_norm = 384.495
INFO (update_parameters): epoch = 9 , avg_abs(trainer_eval-nnue_eval) = 0.999 ,
avg_abs(nnue_eval) = 86 , avg_relative_error = 0.0116163 , batch_size = 1000 ,
rad_norm = 383.212
INFO (update_parameters): epoch = 10 , avg_abs(trainer_eval-nnue_eval) = 0 , av
_abs(nnue_eval) = 68 , avg_relative_error = 0 , batch_size = 1000 , grad_norm =
386.544

PROGRESS (calc_loss): Sun Oct 25 11:42:09 2020, 1000000 sfens, 14627 sfens/seco
d, epoch 10
  - learning rate = 1
  - startpos eval = -90
  - test_cross_entropy_eval = 0.686913
  - test_cross_entropy_win  = 0.689483
  - test_entropy_eval       = 0.242816
  - test_entropy_win        = 0.17987
  - test_cross_entropy      = 0.686913
  - test_entropy            = 0.242816
  - learn_cross_entropy_eval = 0.684886
  - learn_cross_entropy_win  = 0.686634
  - learn_entropy_eval       = 0.24585
  - learn_entropy_win        = 0.179299
  - learn_cross_entropy      = 0.684886
  - learn_entropy            = 0.24585
  - norm = 179998
  - move accuracy = 15.45%
INFO (check_health): layer 0 - HalfKP(Friend)[41024->256x2]
  - observed 36765 (out of 43979) features
  - (min, max) of pre-activations = -0.51018, 1.26212 (limit = 258.008)
  - largest min activation = 0.194051 , smallest max activation = 0.777574
  - avg_abs_bias   = 0.489351
  - avg_abs_weight = 0.0136302
  - clipped 0.165504% of outputs
INFO (check_health): layer 2 - AffineTransform[32<-512]
  - avg_abs_bias        = 0.476143
  - avg_abs_bias_diff   = 9.44856e-05
  - avg_abs_weight      = 0.0369289
  - avg_abs_weight_diff = 4.5695e-05
INFO (check_health): layer 3 - ClippedReLU[32]
  - largest min activation = 0.270169 , smallest max activation = 0.707944
  - clipped 95.3126% of outputs
INFO (check_health): layer 4 - AffineTransform[32<-32]
  - avg_abs_bias        = 0.528737
  - avg_abs_bias_diff   = 0.000241153
  - avg_abs_weight      = 0.146225
  - avg_abs_weight_diff = 0.000136273
INFO (check_health): layer 5 - ClippedReLU[32]
  - largest min activation = 0 , smallest max activation = 0.699718
  - clipped 90.7555% of outputs
INFO (check_health): layer 6 - AffineTransform[1<-32]
  - avg_abs_bias        = 0.152287
  - avg_abs_bias_diff   = 0.0185096
  - avg_abs_weight      = 0.074096
  - avg_abs_weight_diff = 0.000941143
INFO (update_parameters): epoch = 11 , avg_abs(trainer_eval-nnue_eval) = 0.001
 avg_abs(nnue_eval) = 90 , avg_relative_error = 1.11111e-05 , batch_size = 1000
, grad_norm = 381.33
```

### CODE REVIEW

**Comment by nodchip** [comment]
Verbose messages look good to me.  I merged it.
